### PR TITLE
voice_local: fixes and improvements

### PR DESCRIPTION
**Pull request by Fernando-A-Rocha**
This PR doesn't make any new major enhancements to the resource.

It fixes the scripts in terms of logic and security, and adds two important features:

- new meta.xml settings to configure the following:
- max voice distance - default 25 units
- show talking icon above player head - default true
- voice sound boost multiplier - default 6.0

Testing is recommended on a server with at least 1 other player.

### CODE REVIEW

**Comment by Dutchman101** [comment]
The talking icon should remain above any *actively talking* player's head. Even for the local player itself while they are speaking into mic using the voice_ptt bind.

Please keep feature parity, so that means the above, and removing "VOICE: " texts in the bottom of the screen.
Thanks for refactoring!
@Fernando-A-Rocha

**Comment by Fernando-A-Rocha** [comment]
Aww I thought voice text in the bottom of the screen was a good idea. Can I leave it in as optional and disabled by default through meta.xml settings?

**Comment by Dutchman101** [comment]
I've tested and it appears not to be working
Remote players don't hear me, and don't see the speaking icon above my head (while it shows locally). While it works for current version

Your refactored version is bugged still... @Fernando-A-Rocha

**Comment by Fernando-A-Rocha** [comment]
I will perform multiplayer tests again and get back to you. I was using a VM and it's sketchy to get to hear myself... Don't worry, this will be resolved.

**Comment by Fernando-A-Rocha** [comment]
@Dutchman101 Tested. Please give feedback.

Video: https://www.youtube.com/watch?v=ZsMdG7jf-tI

**Comment by Fernando-A-Rocha** [comment]
I added a necessary voice boost (multiplier) set to 6.0.
Voice sound volumes at 1.0 cannot be heard, a boost is required.

File: [gameplay]/voice_local/client.lua [hunk 1]
```diff
@@ -1,155 +1,144 @@
...
 
-function voiceStart(source)
-    if (isElement(source) and getElementType(source) == "player") then
-        if streamedPlayers[source] ~= nil then
-            streamedPlayers[source] = true
-        end
+    if streamedPlayers[source] == nil then
+        setSoundVolume(source, 0)
+        streamedPlayers[source] = false
+        triggerServerEvent("voice_local:addToPlayerBroadcast", resourceRoot, source)
```
Comment (ds1-e) [inline [gameplay]/voice_local/client.lua:103]:
```suggestion
        triggerServerEvent("voice_local:addToPlayerBroadcast", localPlayer, source)
```

File: [gameplay]/voice_local/client.lua [hunk 2]
```diff
@@ -1,155 +1,144 @@
...
 
-function voiceStop(source)
-    if (isElement(source) and getElementType(source) == "player") then
-        if streamedPlayers[source] ~= nil then
-            streamedPlayers[source] = false
-        end
+    if streamedPlayers[source] ~= nil then
+        setSoundVolume(source, 0)
+        streamedPlayers[source] = nil
+        triggerServerEvent("voice_local:removeFromPlayerBroadcast", resourceRoot, source)
```
Comment (ds1-e) [inline [gameplay]/voice_local/client.lua:113]:
```suggestion
        triggerServerEvent("voice_local:removeFromPlayerBroadcast", localPlayer, source)
```

File: [gameplay]/voice_local/client.lua [hunk 3]
```diff
@@ -1,155 +1,144 @@
...
+addEventHandler("onClientResourceStart", resourceRoot, function()
+    for _, player in pairs(getElementsByType("player", root, true)) do
+        if player ~= localPlayer and streamedPlayers[player] == nil then
+            setSoundVolume(player, 0)
+            streamedPlayers[player] = false
         end
     end
-end
-addEventHandler("onClientPlayerJoin", root, playerJoin)
+    triggerServerEvent("voice_local:setPlayerBroadcast", resourceRoot, streamedPlayers)
```
Comment (ds1-e) [inline [gameplay]/voice_local/client.lua:81]:
```suggestion
    triggerServerEvent("voice_local:setPlayerBroadcast", localPlayer, streamedPlayers)
```

File: [gameplay]/voice_local/client.lua [hunk 4]
```diff
@@ -1,155 +1,144 @@
...
+end, false)
 
-function playerQuit()
+-- Handle remote/other player quit
+addEventHandler("onClientPlayerQuit", root, function()
     if streamedPlayers[source] ~= nil then
         streamedPlayers[source] = nil
-        setSoundVolume(source, 0)
-        triggerServerEvent("voice:removePlayerBroadcast", resourceRoot, localPlayer, source)
+        triggerServerEvent("voice_local:removeFromPlayerBroadcast", resourceRoot, source)
```
Comment (ds1-e) [inline [gameplay]/voice_local/client.lua:88]:
```suggestion
        triggerServerEvent("voice_local:removeFromPlayerBroadcast", localPlayer, source)
```

File: [gameplay]/voice_local/server.lua [hunk 5]
```diff
@@ -1,72 +1,128 @@
...
+    if settingData then
+        settings[settingName].value = get("*"..settingData.key)
+    end
+end
 
-    for player, _ in pairs(players) do
-        table.insert(broadcasts[thePlayer], player)
+addEventHandler("onPlayerResourceStart", root, function(res)
+    if res == resource then
+        triggerClientEvent(source, "voice_local:updateSettings", resourceRoot, settings)
```
Comment (ds1-e) [inline [gameplay]/voice_local/server.lua:16]:
```suggestion
        triggerClientEvent(source, "voice_local:updateSettings", source, settings)
```

File: [gameplay]/voice_local/server.lua [hunk 6]
```diff
@@ -1,72 +1,128 @@
...
-        triggerClientEvent(listener, "voice_cl:onClientPlayerVoiceStart", resourceRoot, source)
+addEventHandler("onPlayerVoiceStart", root, function()
+    if not broadcasts[source] then
+        -- Somehow if the system still hasn't loaded the player, prevent them from talking
+        cancelEvent()
+        return
     end
-end
-addEventHandler("onPlayerVoiceStart", root, playerVoiceStart)
+    triggerClientEvent(broadcasts[source], "voice_local:onClientPlayerVoiceStart", resourceRoot, source)
```
Comment (ds1-e) [inline [gameplay]/voice_local/server.lua:113]:
```suggestion
    triggerClientEvent(broadcasts[source], "voice_local:onClientPlayerVoiceStart", source, source)
```

File: [gameplay]/voice_local/server.lua [hunk 7]
```diff
@@ -1,72 +1,128 @@
...
-    for _, listener in ipairs(broadcasts[source]) do
-        triggerClientEvent(listener, "voice_cl:onClientPlayerVoiceStop", resourceRoot, source)
+addEventHandler("onPlayerVoiceStop", root, function()
+    if not broadcasts[source] then
+        return
     end
-end
-addEventHandler("onPlayerVoiceStop", root, playerVoiceStop)
\ No newline at end of file
+    triggerClientEvent(broadcasts[source], "voice_local:onClientPlayerVoiceStop", resourceRoot, source)
```
Comment (ds1-e) [inline [gameplay]/voice_local/server.lua:120]:
```suggestion
    triggerClientEvent(broadcasts[source], "voice_local:onClientPlayerVoiceStop", source, source)
```

File: [gameplay]/voice_local/server.lua [hunk 8]
```diff
@@ -1,72 +1,128 @@
...
+    local opx, opy, opz = getElementPosition(otherPlayer)
+    return getDistanceBetweenPoints3D(px, py, pz, opx, opy, opz) <= maxDist
 end
-addEvent("voice:setPlayerBroadcast", true)
-addEventHandler("voice:setPlayerBroadcast", resourceRoot, setPlayerBroadcast)
 
-function addToPlayerBroadcast(thePlayer, player)
-    if not broadcasts[thePlayer] then
-        broadcasts[thePlayer] = {}
+addEventHandler("voice_local:setPlayerBroadcast", resourceRoot, function(players)
```
Comment (ds1-e) [inline [gameplay]/voice_local/server.lua:48]:
```suggestion
addEventHandler("voice_local:setPlayerBroadcast", root, function(players)
```

File: [gameplay]/voice_local/server.lua [hunk 9]
```diff
@@ -1,72 +1,128 @@
...
     end
+    setPlayerVoiceBroadcastTo(client, broadcasts[client])
+end, false)
 
-    table.insert(broadcasts[thePlayer], player)
-    setPlayerVoiceBroadcastTo(thePlayer, broadcasts[thePlayer])
-end
-addEvent("voice:addToPlayerBroadcast", true)
-addEventHandler("voice:addToPlayerBroadcast", resourceRoot, addToPlayerBroadcast)
+addEventHandler("voice_local:addToPlayerBroadcast", resourceRoot, function(player)
```
Comment (ds1-e) [inline [gameplay]/voice_local/server.lua:65]:
```suggestion
addEventHandler("voice_local:addToPlayerBroadcast", root, function(player)
```

File: [gameplay]/voice_local/server.lua [hunk 10]
```diff
@@ -1,72 +1,128 @@
...
 
-    setPlayerVoiceBroadcastTo(thePlayer, broadcasts[thePlayer])
-end
-addEvent("voice:removePlayerBroadcasts", true)
-addEventHandler("voice:removePlayerBroadcasts", resourceRoot, removePlayerBroadcasts)
+    table.insert(broadcasts[client], player)
+    setPlayerVoiceBroadcastTo(client, broadcasts[client])
+end, false)
+
+addEventHandler("voice_local:removeFromPlayerBroadcast", resourceRoot, function(player)
```
Comment (ds1-e) [inline [gameplay]/voice_local/server.lua:89]:
```suggestion
addEventHandler("voice_local:removeFromPlayerBroadcast", root, function(player)
```

File: [gameplay]/voice_local/client.lua [hunk 11]
```diff
@@ -1,155 +1,144 @@
...
-end
-addEvent("voice_cl:onClientPlayerVoiceStop", true)
-addEventHandler("voice_cl:onClientPlayerVoiceStop", resourceRoot, voiceStop)
+end)
 
-function table.empty(a)
-    if type(a) ~= "table" then
-        return false
+-- Update player talking status (for displaying)
+addEventHandler("voice_local:onClientPlayerVoiceStart", resourceRoot, function(player)
```
Comment (ds1-e) [inline [gameplay]/voice_local/client.lua:118]:
```suggestion
addEventHandler("voice_local:onClientPlayerVoiceStart", root, function(player)
```

File: [gameplay]/voice_local/client.lua [hunk 12]
```diff
@@ -1,155 +1,144 @@
...
+    if not (isElement(player) and getElementType(player) == "player") then return end
+
+    if player == localPlayer then
+        localPlayerTalking = true
+    elseif streamedPlayers[player] ~= nil then
+        streamedPlayers[player] = true
+    end
+end)
+addEventHandler("voice_local:onClientPlayerVoiceStop", resourceRoot, function(player)
```
Comment (ds1-e) [inline [gameplay]/voice_local/client.lua:127]:
```suggestion
addEventHandler("voice_local:onClientPlayerVoiceStop", root, function(player)
```

File: [gameplay]/voice_local/client.lua [hunk 13]
```diff
@@ -1,155 +1,144 @@
...
+        localPlayerTalking = false
+    elseif streamedPlayers[player] ~= nil then
+        streamedPlayers[player] = false
     end
+end)
 
-    return next(a) == nil
-end
\ No newline at end of file
+addEventHandler("voice_local:updateSettings", resourceRoot, function(settingsFromServer)
```
Comment (ds1-e) [inline [gameplay]/voice_local/client.lua:137]:
```suggestion
addEventHandler("voice_local:updateSettings", localPlayer, function(settingsFromServer)
```

In this case event is sent only to localPlayer

File: [gameplay]/voice_local/client.lua (hunk 3 above)
Comment (ds1-e) [reply [gameplay]/voice_local/client.lua:81]: Simply use localPlayer

**Review by ds1-e — CHANGES_REQUESTED** [review]
Some smaller changes

**Comment by Fernando-A-Rocha** [comment]
I've applied your event handler source suggestions @ds1-e

**Review by ds1-e — APPROVED** [review]

**Comment by Dutchman101** [comment]
Thanks for addressing everything, it works fine after testing too.

**Comment by Fernando-A-Rocha** [comment]
> Thanks for addressing everything, it works fine after testing too.

I'm glad!

Guys, please suggest new features for this resource. I will happily make a new PR.

Reminder: it's a default local 3D voice system, intended to be used by any server, suitable for roleplay, freeroam etc.

Some ideas:
- add conditions, functions and events to determine when a player can't talk (mute), e.g. when dead or knocked out (RP servers)
- add customizable on-screen text/icons to display when the voice is being used, like many games do, aside from the icon above player head which can be unwanted (right now it can be disabled)
- shout and whisper modes
- voice effects 
- customizable radio channels or player-to-player voice to simulate phone calls

**Comment by Dutchman101** [comment]
@Fernando-A-Rocha
The resource still suffers from a bug that was present in the initial version, that you can sometimes hear random player's voices (when no one is near you; those players being on the other side of the map). I got reports about it again, and then witnessed a case of it occuring.

**Comment by Fernando-A-Rocha** [comment]
Can someone please review the code to get more opinions on why the situation described by @Dutchman101 happens?

**Pull request merged**